src/rpc/blockchain.cpp

```diff
@@ -162,16 +162,21 @@ UniValue blockToJSON(const CBlock& block, const CBlockIndex* tip, const CBlockIn
     result.pushKV("versionHex", strprintf("%08x", block.nVersion));
     result.pushKV("merkleroot", block.hashMerkleRoot.GetHex());
     UniValue txs(UniValue::VARR);
-    for(const auto& tx : block.vtx)
-    {
-        if(txDetails)
-        {
+    if (txDetails) {
+        CBlockUndo blockUndo;
+        const bool have_undo = !IsBlockPruned(blockindex) && UndoReadFromDisk(blockUndo, blockindex);
```

Comment:
Great idea!

Yes, according to https://github.com/bitcoin/bitcoin/blob/1b313cacc99a1b372238f9036abed5491f9d28f7/src/validation.h#L370 it looks like block and undo data are always deleted together.

I definitely might be missing something, but from a quick search I couldn't find any other reason we'd delete `rev` but not `blk` files.

Reply:
I think it's not unreasonable to delete undo data while keeping block data (that means you could still serve it to other syncing nodes, and rescan, but can't reorg out of it). That just was never implemented.

Reply:
OK, you both make some great points. I can think of three approaches at the moment:

- Use `GetUndoChecked` which means if someone deletes undo data in the future, they'll cause this RPC to fail
- Silently leave out the fee if block data is present but undo data is missing
- Add some verbose documentation that says "we'll include the fee if we have the data needed to efficiently calculate it"

Do any of these approaches seem reasonable to both of you?

Reply:
As @sipa says it's not implemented and I think there is no one working on something like that (Side-note: I will look into that now :) ). So I would say the first option would still be my first choice if we could have a test that would reliably fail when the change comes. But that seems unrealistic, at least if it is implemented the way I envision it now and generally it would be poking around in the dark. So, I guess I would be happy if you add some docs but you don't have to since with the current state of the code the fee should be there reliably.

Reply:
I think either leaving it like it is now, or returning an error when the undo file is not available but the block file is (since that's currently an unexpected condition) are both fine.